### PR TITLE
peer: Separate ping ticker functionality from outHandler.

### PR DESCRIPTION
**Pull request by jongillham**

### CODE REVIEW

**Comment by jongillham** [comment]
Requires pull request #626.

File: server.go [hunk 1]
```diff
@@ -1573,9 +1568,7 @@ func (s *server) establishConn(sp *serverPeer) error {
 	if err != nil {
 		return err
 	}
-	if err := sp.Connect(conn); err != nil {
-		return err
-	}
+	sp.Connect(conn)
```
Comment (dajohi) [inline server.go]: If the connect fails, you will get both:
log.Errorf("Cannot start peer %v: %v", p, err)
srvrLog.Debugf("Connected to %s", sp.Addr())

Comment (jongillham) [inline server.go]: @dajohi thanks for the review. I'll update with a solution that gives non-contradictory logging.

**Comment by jongillham** [comment]
@davecgh I thought I would start back with the peer refactoring. Better separation of concerns is the only benefit of this request - at the expense of more state variables. However it helps work towards removing the stalling handler polling and potentially remove the need for code to keep track of 'pending peers'.

**Comment by davecgh** [comment]
@jongillham Sounds good.  Both this PR and #629 are almost to the top of me PR review list, so hopefully I should be able to get to them before too much longer.

**Comment by jrick** [comment]
I agree with making it a new goroutine, as @davecgh suggested.  reads better and is easier to understand.

**Comment by jongillham** [comment]
@davecgh it certainly reads easier to have a separate goroutine for ping ticker. That is how the pull request originally was however I remember having a discussion with you on another thread about the desire to keep the number of goroutines for each peer down to a minimum.

After that I had a quick look at the Go internals to see how `time.AfterFunc` worked and got the impression that it amortised its cost by mainly creating one goroutine to handle all firing of timer events. Here's a [link](https://github.com/golang/go/blob/2481481ff798636907376bfdf7e8c7558b8b930e/src/runtime/time.go#L64) to the code I looked at.

Of course, if I misread the `time` code and one goroutine is created for each `time.AfterFunc` and/or you don't have the requirement for keeping the number of goroutines down to a minimum then I would certainly vote for increased readability as you suggest.

**Comment by davecgh** [comment]
@jongillham: That's a fair point about amortization and one I didn't actually know, although it does make sense the stdlib would do that.  I do recall having the discussion about goroutines (although I'd have to review the specific details), and I definitely think it's something we want to keep an eye on.

At any rate, I think given the readability improvement, and the fact that several people prefer the goroutine for this, it justifies the addition.  Sorry for the mixed signals on that front.

If you want to update it to the goroutine approach, I'll go ahead and merge it after that.  I'll be looking at your other peer PR (#629) next.

**Comment by jongillham** [comment]
@davecgh updated.

**Pull request merged**